sycl/doc/extensions/proposed/sycl_ext_oneapi_record_event.asciidoc

```diff
@@ -0,0 +1,210 @@
+link:../proposed/sycl_ext_oneapi_enqueue_functions.asciidoc[
+sycl_ext_oneapi_enqueue_functions]
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_RECORD_EVENT` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New device aspect
+
+This extension adds the `ext_oneapi_queue_event_recording` enumerator to the
+`sycl::aspect` enumeration.
+
+```
+namespace sycl {
+
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_queue_event_recording
+};
+
+} // namespace sycl
+```
+
+When a device has this aspect, the `record_event` function may be called for a
+queue on this device even if the queue is not constructed with the property
+`property::queue::enable_profiling`.
+
+=== New free function
+
+This extension adds the following free function.
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+namespace sycl::ext::oneapi::experimental {
+
+event record_event(const queue& q);
+
+} // namespace sycl::ext::oneapi::experimental
+----
+!====
+
+_Effects:_ Enqueues a command barrier to `q`.
+Any commands submitted after this barrier cannot begin execution until all
+previously submitted commands have completed.
+
+_Returns:_ An event which represents the completion of the barrier.
+The event's status becomes `info::event_command_status::complete` when all
+commands submitted to the queue prior to the call to `record_event` have
+completed.
+The event's `info::event_profiling::command_submit` timestamp reflects the
+time at which `record_event` is called.
+The event's `info::event_profiling::command_end` timestamp reflects the time
+at which the event enters the "complete" state.
+The event's `info::event_profiling::command_start` timestamps is the same as
+the `info::event_profiling::command_end` timestamp.
+
+_Throws:_ A synchronous `exception` with the `errc::invalid` error code if the
+queue was not constructed with the `property::queue::enable_profiling` property
+and if the queue's device does not have the aspect
+`ext_oneapi_queue_event_recording`.
+
+[_Note:_ In order to understand why the "start" and "end" timestamps are the
+same, think of the barrier as an empty kernel with an implicit set of
+dependencies on all previous commands in the same queue.
+This theoretical kernel starts executing when the dependencies are resolved.
+Since the kernel is empty, the end time is the same as the start time.
+_{endnote}_]
```

Comment:
I'm not sure, but I'm wondering if we should say that the `command_start` and `command_end` timestamps _are_ the same, or if we should say something slightly weaker like they're _approximately the same_.

The case I'm thinking about is one where an implementation chooses to implement this as an empty kernel with a dependency list, but that implementation isn't smart enough to optimize away the empty kernel.  That may result in some negligible difference between the start and end times.

If we can't think of a use-case where a user would need to be able to check that the start times and end times are exactly the same, weakening this to reduce the burden on implementers might be a good idea. Computing elapsed time the way you do in the example should give the same answer even if a negligible difference was allowed.

Reply:
I don't feel strongly about this, so we could weaken the language here.  If we do that, I probably would change the example, though to be:

```
uint64_t elapsed =
    end.get_profiling_info<sycl::info::event_profiling::command_end>() -
    start.get_profiling_info<sycl::info::event_profiling::command_end>();
```

If the spec says that the start and end times might be different, the canonical way to compute the elapsed time should exclude the time it takes to "execute" the barrier, even if it will typically be small.

Reply:
> If the spec says that the start and end times might be different, the canonical way to compute the elapsed time should exclude the time it takes to "execute" the barrier, even if it will typically be small.

I agree, but I think this was already true the way you'd written the spec before:

```c++
  uint64_t elapsed =
    end.get_profiling_info<sycl::info::event_profiling::command_start>() -
    start.get_profiling_info<sycl::info::event_profiling::command_end>();
  std::cout << "Execution time: " << elapsed << " (nanoseconds)\n";
```

The start of the end event is when the second barrier started, and the end of the start event is when the first barrier finished.  So, the time between these two events doesn't include the time taken to "execute" either barrier.

With your new suggestion:
```c++
uint64_t elapsed =
    end.get_profiling_info<sycl::info::event_profiling::command_end>() -
    start.get_profiling_info<sycl::info::event_profiling::command_end>();
```
Using the end of the end event here now means you're including the time it takes to "execute" the barrier.

...I think.

Reply:
Doh!  I got the "start" and "end" confused.  Yes, you're right.  The canonical way to measure the elapsed time should be the way it is already in the example.

Reply:
Wording relaxed in b36c0ffe7902488cfdb153edabbf075711231cc4.